### DISCUSSION TITLE
Brave bookmarks recursively

### DISCUSSION THREAD

**Discussion by Alarid** [johnlindquist/kit#136] [Share]
Hey there 👋 

I just discovered Script Kit today, and I absolutely love it, awesome idea @johnlindquist 🔥

I made a few changes to your script `chrome-bookmarks` so I can use it with Brave (small change), but also to parse the bookmarks JSON recursively. This is particularly useful to me as I nest all my bookmarks in subfolders of the bookmark bar. With this change, I can open any bookmark (however deep it is nested in the bookmarks JSON). 

I thought I'd share the script here, in case anyone find it useful too.

```javascript
// Menu: Brave Bookmarks
// Description: Select and open a bookmark from Brave

function parseBookmarks(node) {
  const bookmarks = [];
  node.forEach((item) => {
    if (item.type === "url") {
      bookmarks.push(item);
    } else if (item.type === "folder") {
      bookmarks.push(...parseBookmarks(item.children));
    }
  });
  return bookmarks;
}

let bookmarksFile = await readFile(
  home(
    "Library/Application Support/BraveSoftware/Brave-Browser/Default/Bookmarks"
  )
);
const rootNode = JSON.parse(bookmarksFile);
const bookmarkBar = rootNode.roots.bookmark_bar;
const bookmarks = parseBookmarks(bookmarkBar.children);

let url = await arg(
  "Select bookmark",
  bookmarks.map(({ name, url }) => {
    return {
      name: name || url,
      description: url,
      value: url,
    };
  })
);

exec(`open ${url}`);

```

**Comment by johnlindquist** [comment]
Awesome, thanks for sharing!

One of the goals of Script Kit is to enable people to easily tweak someone else's script to fit their personal needs. So I'm very glad this worked out for you 😊

**Comment by cezary-draus** [comment]
Hi @Alarid that is so cool! Today I also tried Script Kit for the first time and also noticed a problem with folders in bookmarks. I solved it with a flat map (my first try in type script):
```typescript
import "@johnlindquist/kit"
    
// Menu: Chrome Bookmarks
// Description: Select and open a bookmark from Chrome
// Author: John Lindquist, Cezary Draus
// Twitter: @johnlindquist, @CezaryDraus

let bookmarks = await readFile(
  home(
    "Library/Application Support/Google/Chrome/Default/Bookmarks"
  )
)

bookmarks = JSON.parse(bookmarks)
bookmarks = bookmarks.roots.bookmark_bar.children
console.log(bookmarks);

let url = await arg(
  "Select bookmark",
  bookmarks.flatMap(element => {
    if (Array.isArray(element.children)) {
      return element.children
    } else {
      return [element]
    }
  }).map(({ name, url }) => {

    return {
      name,
      description: url,
      value: url,
    }
  })
)

exec(`open "${url}"`)
```

**Comment by cezary-draus** [comment]
@johnlindquist How can we upgrade your browser bookmarks script? Or its non-recursive by design, so you can fish us :)?

**Reply by johnlindquist** [reply]
@cezary-draus Not sure you meant to ping me? Is there something I can help with? That's not my script...

**Reply by cezary-draus** [reply]
Ah, the script is under your name. That's why I thought it was yours. Anyway, I thought maybe it's worth upgrading it, so it works when having directories. WDYT?
<img width="322" alt="image" src="https://user-images.githubusercontent.com/8237929/210652758-224b6118-9e53-4b00-90a6-dfba8606ac56.png">
